[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Jibby2k1/SPS_Curriculum/blob/main/Intro_Math/Analysis/Independence.ipynb)


Content for and by IEEE Signal Processing Society. (Raul Valle & Contributors)

# Independence

> ⚠️ **Draft — pending instructor review.** The visuals below execute, but execution cannot verify proofs. An SPS instructor should vet the arguments in this notebook before it is taught. Remove this banner after review.

The final workshop of the Analysis track. Independence is the assumption hiding inside every "i.i.d." in every ML paper; Borel–Cantelli tells you which coincidences recur forever; and the laws of large numbers are the *theorems* that make sample averages — Monte Carlo, training loss, every benchmark in this curriculum — mean anything at all.

### Visual setup & helpers
*(Safe to re-run anytime.)*

In [1]:
import numpy as np
import matplotlib.pyplot as plt
rng = np.random.default_rng(0)

---
### 🕐 Session 1 of 2 — *Independence & Borel–Cantelli* (~35 min)
**Goal:** define independence for events, σ-algebras, and random variables; prove both Borel–Cantelli lemmas.
**Builds on:** [Random Variables](./Random_Variables.ipynb). &nbsp; **Feeds into:** Session 2 (laws of large numbers).

---

<details>
<summary>🎓 <b>Teacher notes — Session 1: Independence & Borel–Cantelli</b></summary>

**Timing (~35 min).** 8 min independence and the pairwise trap · 8 min the general definition · 12 min both Borel–Cantelli proofs · 7 min the demo.

**Board first — independence as an information statement.** Learning that $A$ happened tells you *nothing* about $B$. The multiplication rule is the algebraic fingerprint: conditioning on $A$ divides by $P(A)$, and independence is exactly the condition that leaves $P(B)$ unchanged. Define it that way and the formula becomes a consequence rather than a definition to memorise.

**Then the trap, and give the counterexample in full.** Independence for many events requires the product rule for **every** sub-collection, not just the whole batch. Two fair flips with $A$ = first heads, $B$ = second heads, $C$ = the flips agree: any *two* are independent, and all three are not, since $P(A\cap B\cap C) = 1/4 \ne 1/8$. Have the room verify each pair before revealing the triple — the surprise only lands if they have checked the pairs themselves.

**The general definition is worth stating even if you do not dwell on it.** Random variables are independent when their generated σ-algebras are, equivalently when the joint law is the **product measure**. Then Fubini factors expectations: $E[XY] = E[X]E[Y]$. Say the one-way implication loudly — **independent implies uncorrelated, never the converse** — and point at [Information Theory](../Information_Theory/Information_Theory.ipynb)'s $y = x^2$ example, where correlation is zero and dependence is total.

**Borel–Cantelli deserves to be framed as a dichotomy about coincidences.** Do rare events recur forever, or eventually stop? The answer depends only on whether $\sum P(A_n)$ converges. Both proofs are short and both should be done.

**Emphasise the asymmetry in the hypotheses — it is the memorable part.** BC-I needs **no independence at all**: it is just subadditivity plus the tail of a convergent series. BC-II needs independence, and genuinely so. Ask for a counterexample without it: take $A_n = A$ for a single event with $0 < P(A) < 1$. Then $\sum P(A_n) = \infty$, but the i.o. event is just $A$, which has probability less than 1. One line, and it shows exactly what independence is buying.

**Point at the $1-x \le e^{-x}$ step in BC-II.** That inequality is what converts a product into an exponential of a sum, so a divergent sum forces the product to zero. Small trick, load-bearing.

**Close on the demo's honest limitation.** The plot shows one realisation each of $p_n = 1/n^2$ and $p_n = 1/n$. It illustrates the dichotomy; it cannot verify an "infinitely often" statement from 5000 samples. Say so — the theorem is about the infinite tail, and no finite simulation reaches it.
</details>

## Independence

💡 **Intuition.** Independence means *information about one tells you nothing about the other*: learning that $A$ happened doesn't move the odds of $B$. The multiplication rule $P(A \cap B) = P(A)P(B)$ is the algebraic fingerprint of that idea — conditioning on $A$ (dividing by $P(A)$) leaves $P(B)$ untouched. The subtlety is that for many events, independence must hold for **every sub-collection**, not just the whole batch.

**Definition.** Events $A_1, \dots, A_n$ are *independent* if for every subset of indices $\{i_1 < \cdots < i_k\}$:

$$P(A_{i_1} \cap \cdots \cap A_{i_k}) = P(A_{i_1}) \cdots P(A_{i_k}).$$

Pairwise independence is strictly weaker (classic counterexample: two fair coin flips, with $A$ = first is heads, $B$ = second is heads, $C$ = the flips agree — any two are independent, all three are not: $P(A \cap B \cap C) = \tfrac14 \ne \tfrac18$).

**Definition (the general form).** σ-algebras $\mathcal{G}_1, \dots, \mathcal{G}_n$ are independent if any choice of one event from each is independent. Random variables $X_1, \dots, X_n$ are independent if their generated σ-algebras $\sigma(X_i) = \{X_i^{-1}(B)\}$ are — equivalently, the joint law is the **product measure**: $P_{(X_1, \dots, X_n)} = P_{X_1} \otimes \cdots \otimes P_{X_n}$. Fubini's theorem then factors expectations: $E[X Y] = E[X]E[Y]$ for independent $X, Y \in L^1$ — so independent variables are uncorrelated (never the converse!).

## Borel–Cantelli

For events $A_1, A_2, \dots$, define $\{A_n \text{ i.o.}\} = \limsup_n A_n = \bigcap_{n=1}^{\infty} \bigcup_{k \ge n} A_k$ — the outcomes lying in *infinitely many* $A_n$ ("infinitely often"). Note the shape: it's the event version of $\limsup$ from [Sequences & Series](./Numerical_Sequences_and_Series.ipynb).

💡 **Intuition.** Borel–Cantelli is a zero–one dichotomy about recurring coincidences. If the probabilities $P(A_n)$ shrink fast enough to have a *finite sum*, then with probability 1 only finitely many happen — the tail of a convergent series is too small to keep producing events. If the sum *diverges* **and** the events are independent, they happen infinitely often with probability 1 — infinite total probability with no coordination cannot be dodged forever.

### Proof: Borel–Cantelli I

If $\sum_n P(A_n) < \infty$, then $P(A_n \text{ i.o.}) = 0$.

For every $n$, $\{A_k \text{ i.o.}\} \subseteq \bigcup_{k \ge n} A_k$, so by subadditivity

$$P(A_k \text{ i.o.}) \le P\Big(\bigcup_{k \ge n} A_k\Big) \le \sum_{k \ge n} P(A_k) \xrightarrow{n \to \infty} 0,$$

the tail of a convergent series. $\blacksquare$ *(No independence needed.)*

### Proof: Borel–Cantelli II

If the $A_n$ are independent and $\sum_n P(A_n) = \infty$, then $P(A_n \text{ i.o.}) = 1$.

It suffices to show $P\big(\bigcap_{k \ge n} A_k^c\big) = 0$ for each $n$ (then the union over $n$ of these null sets is null, and the complement — the i.o. event — has probability 1). Using independence of complements and $1 - x \le e^{-x}$:

$$P\Big(\bigcap_{k=n}^{m} A_k^c\Big) = \prod_{k=n}^{m} (1 - P(A_k)) \le \exp\Big(-\sum_{k=n}^{m} P(A_k)\Big) \xrightarrow{m \to \infty} 0,$$

since the exponent diverges to $-\infty$. Continuity from above finishes it. $\blacksquare$

In [2]:
# Borel–Cantelli in action: A_n = {U_n < p_n} for independent uniforms
# Σ 1/n² < ∞  → events stop;   Σ 1/n = ∞ → they never stop
N = 5000
U = rng.random(N)
nn = np.arange(1, N + 1)

fig, axes = plt.subplots(1, 2, figsize=(9.5, 2.7), sharey=True)
for ax, p, title in [(axes[0], 1/nn**2, "$p_n = 1/n^2$ (Σ<∞): finitely many hits"),
                     (axes[1], 1/nn,    "$p_n = 1/n$ (Σ=∞): hits keep coming")]:
    hits = nn[U < p]
    ax.eventplot(hits, linelengths=0.8)
    ax.set_title(f"{title} — {len(hits)} hits in {N}")
    ax.set_xlabel("n")
plt.tight_layout(); plt.show()

/tmp/ipykernel_1900545/334830335.py:14: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.tight_layout(); plt.show()


**What just happened.** Two event streams with the *same* mechanism and opposite long-run behaviour. With $p_n = 1/n^2$ the hits cluster at small $n$ and then stop entirely. With $p_n = 1/n$ they keep arriving all the way to $n = 5000$, thinning out but never ceasing.

**And the only thing that differs is whether a series converges.** $\sum 1/n^2 = \pi^2/6 < \infty$; $\sum 1/n$ diverges. Borel–Cantelli turns that purely analytic fact into a statement about which coincidences recur forever — which is a genuinely surprising bridge between [Sequences & Series](./Numerical_Sequences_and_Series.ipynb) and probability.

**The asymmetry in the two lemmas is the part worth remembering.** BC-I — convergent sum implies finitely many events — needs **no independence whatsoever**. It is subadditivity plus the tail of a convergent series, and it holds for arbitrarily entangled events. BC-II — divergent sum implies infinitely many — *requires* independence, and genuinely so.

The counterexample takes one line: let every $A_n$ be the *same* event $A$ with $0 < P(A) < 1$. Then $\sum P(A_n) = \infty$, but $\{A_n \text{ i.o.}\}$ is just $A$ itself, which has probability strictly less than 1. Independence is what prevents events from conspiring to fail together, and without it a divergent sum guarantees nothing.

**Note the trick that makes BC-II work.** The inequality $1 - x \le e^{-x}$ converts $\prod(1 - P(A_k))$ into $\exp(-\sum P(A_k))$, so a divergent sum drives the product to zero. Small step, load-bearing — it is what turns "infinite total probability" into "cannot be dodged."

**One honest limitation of this plot.** It shows a single realisation of each stream over 5000 steps. That *illustrates* the dichotomy; it cannot verify it. "Infinitely often" is a statement about the infinite tail, and no finite simulation reaches the tail — the $1/n^2$ stream could in principle produce another hit at $n = 10^6$, and the theorem still allows finitely many. What the picture gives is intuition for why the two cases feel different; the proofs give the certainty.

**Where this gets used.** Borel–Cantelli I is the engine inside the strong law of large numbers (Session 2) and inside most almost-sure convergence proofs in probability. The pattern is always the same: bound the probability of a bad event at step $n$, show the bounds are summable, conclude that only finitely many bad events occur — so eventually, and forever after, things go right.

---
### 🕐 Session 2 of 2 — *Laws of Large Numbers* (~40 min)
**Goal:** prove the weak law with Chebyshev; state the strong law; see exactly what they license.
**Builds on:** Session 1.

---

<details>
<summary>🎓 <b>Teacher notes — Session 2: Laws of Large Numbers</b></summary>

**Timing (~40 min).** 10 min the WLLN proof · 8 min weak vs strong · 10 min the two demos · 12 min what breaks without independence.

**Board first — answer "why do averages settle down?" mechanically.** Independent errors *cancel*. Variances of independent variables add, so $\mathrm{Var}(\bar X_n) = \sigma^2/n$ — the average's spread shrinks like $1/\sqrt n$. Chebyshev then converts "tiny variance" into "probably close to the mean," and that is the entire proof. Three lines, and it underwrites every error bar in the curriculum.

**Make the weak/strong distinction concrete, because it is genuinely subtle.** WLLN: for each *fixed large* $n$, $\bar X_n$ is probably close to $\mu$. SLLN: for almost every *run of the experiment*, the whole trajectory converges. The difference is a statement about marginals versus a statement about paths. A useful framing: WLLN allows $\bar X_n$ to wander far infinitely often (just rarely at any fixed $n$); SLLN forbids it. Point at the demo — the plot shows *trajectories*, so it is illustrating the strong law.

**Both demos are verifications, so treat them as such.** The $\sigma/\sqrt n$ table matches theory to three decimals across four orders of magnitude in $n$ — that is the WLLN rate confirmed, not merely illustrated. Have the room predict each row before revealing it; $\sigma = 1$ for Exponential(1) makes the arithmetic trivial and the agreement striking.

**The correlation demo is the most practically valuable cell in the workshop.** Same marginal variance, same 5000 samples, and the standard error is **6.2× larger** for AR(1) with $\rho = 0.95$. Convert that into the number people actually use: effective sample size $n_{\text{eff}} = n(1-\rho)/(1+\rho) = 128$. **Five thousand correlated samples are worth about 128 independent ones — 2.6%.** Ask the room to predict the ratio from that formula before running; $\sqrt{5000/128} = 6.24$ against a measured 6.23 is a satisfying hit.

**Then name where this bites.** MCMC (the [spectral gap](../Stochastic_Processes/Stochastic_Processes_2.ipynb) sets $\rho$, so slow mixing means tiny $n_{\text{eff}}$), time-series cross-validation, and A/B tests with repeated measures on the same users. In all three, treating correlated observations as independent produces error bars that are too small by a factor you can now compute. That is the most common statistical error in applied work, and this cell quantifies it.

**Close the track.** This is the last workshop of the Analysis arc: ℝ exists → its topology → limits work → sets can be measured → probability is a measure → **averages converge**. Every empirical mean in this curriculum now stands on proved ground, and it is worth taking thirty seconds to say so.
</details>

## The Weak Law of Large Numbers

💡 **Intuition.** Why do averages settle down? Independent errors *cancel*: the variance of a sum of independent variables adds, so the variance of the **average** shrinks like $1/n$. Chebyshev converts "tiny variance" into "probably close to the mean" — and that's the whole proof. Every Monte Carlo error bar, every "$\pm$" on a benchmark, is this argument.

### Proof: WLLN (finite-variance version)

Let $X_1, X_2, \dots$ be i.i.d. with mean $\mu$ and variance $\sigma^2 < \infty$, and $\bar{X}_n = \frac{1}{n}\sum_{i=1}^n X_i$. Then for every $\varepsilon > 0$, $P(|\bar{X}_n - \mu| \ge \varepsilon) \to 0$.

Linearity gives $E[\bar{X}_n] = \mu$. Independence makes variances add:

$$\mathrm{Var}(\bar{X}_n) = \frac{1}{n^2} \sum_{i=1}^{n} \mathrm{Var}(X_i) = \frac{\sigma^2}{n}.$$

Chebyshev ([Random Variables](./Random_Variables.ipynb)) then gives

$$P(|\bar{X}_n - \mu| \ge \varepsilon) \le \frac{\sigma^2}{n \varepsilon^2} \xrightarrow{n \to \infty} 0. \;\blacksquare$$

## The Strong Law

**Theorem (SLLN, Kolmogorov).** For i.i.d. $X_i$ with $E[|X_1|] < \infty$:

$$P\Big( \lim_{n \to \infty} \bar{X}_n = \mu \Big) = 1.$$

*(Proof omitted — the standard route runs through Borel–Cantelli I plus a truncation argument; see Durrett §2.4. A pleasant exercise: prove it yourself under the stronger assumption $E[X_1^4] < \infty$, using Chebyshev on fourth moments + Borel–Cantelli I.)*

The difference is real: WLLN says each *fixed large* $n$ is probably fine; SLLN says the *trajectory* $\bar{X}_n(\omega)$ converges for almost every run of the experiment — "almost surely," i.e. off a null set from [Measure Theory](./Measure_Theory.ipynb).

In [3]:
# Watch trajectories converge (SLLN): 20 independent runs of a cumulative average
n_steps, n_paths = 20_000, 20
X = rng.exponential(1.0, size=(n_paths, n_steps))          # μ = 1
avg = np.cumsum(X, axis=1) / np.arange(1, n_steps + 1)

plt.figure(figsize=(8.5, 3))
plt.plot(avg.T, linewidth=0.6, alpha=0.6)
plt.axhline(1.0, color="k", linewidth=1.2)
plt.xscale("log"); plt.ylim(0.4, 1.8)
plt.title("20 sample paths of $\\bar{X}_n$, Exponential(1): every path → μ = 1")
plt.xlabel("n (log scale)")
plt.tight_layout(); plt.show()

/tmp/ipykernel_1900545/772346111.py:12: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.tight_layout(); plt.show()


**What just happened.** Twenty independent runs, each showing the cumulative average of Exponential(1) samples. Early on they wander widely; by $n = 10^4$ every path has collapsed onto $\mu = 1$.

**Notice what is being plotted: trajectories, not marginals.** That is the distinction between the two laws, and it is the reason this demo illustrates the *strong* law rather than the weak one.

- **WLLN** says: for each *fixed large* $n$, $\bar X_n$ is probably close to $\mu$. A statement about the distribution at one time.
- **SLLN** says: for almost every *run of the experiment*, the entire trajectory converges to $\mu$. A statement about paths.

The gap between them is real. The weak law permits a path to wander far from $\mu$ infinitely often, provided it is rare at any given $n$; the strong law forbids that. What you see here — each individual line settling and staying settled — is the strong law's content, and it is what licenses the everyday practice of running one long simulation and trusting its average.

**The log-scaled $x$-axis is doing useful work.** Convergence looks steady across it because the spread shrinks like $1/\sqrt n$, so each *decade* of samples buys a fixed factor in precision — roughly $\sqrt{10} \approx 3.2$. Ten times the compute for three times the accuracy is the fundamental exchange rate of Monte Carlo, and it is why high-precision simulation is expensive.

**Note also that the exponential distribution is skewed**, with occasional large values, and the paths converge anyway. The SLLN needs only $E|X_1| < \infty$ — no variance assumption, no symmetry, no bounded support. It is a remarkably cheap hypothesis for a remarkably strong conclusion.

That cheapness has a limit worth flagging: for a Cauchy distribution $E|X_1| = \infty$, and the sample average famously does **not** converge — it has the same distribution as a single sample, no matter how many you take. Averaging is not universally stabilising; it is stabilising when the mean exists.

**Where the strong law comes from.** The standard proof route runs through Borel–Cantelli I from Session 1: bound the probability that $|\bar X_n - \mu|$ exceeds $\varepsilon$, show those bounds are summable, and conclude that only finitely many excursions occur — so eventually, and forever after, the path stays close. The two sessions of this workshop are one argument.

In [4]:
# And the WLLN rate: spread of X̄_n shrinks like 1/√n (variance like 1/n)
ns = np.array([10, 100, 1000, 10000])
trials = np.array([[rng.exponential(1.0, n).mean() for _ in range(2000)] for n in ns])

print(" n      std of X̄_n   σ/√n prediction")
for n_val, row in zip(ns, trials):
    print(f"{n_val:>6}   {row.std():.4f}       {1/np.sqrt(n_val):.4f}")

 n      std of X̄_n   σ/√n prediction
    10   0.3109       0.3162
   100   0.0996       0.1000
  1000   0.0319       0.0316
 10000   0.0100       0.0100


**What just happened.** The WLLN's *rate*, verified to three decimals across four orders of magnitude:

| $n$ | measured std | $\sigma/\sqrt n$ |
|---|---|---|
| 10 | 0.3109 | 0.3162 |
| 100 | 0.0996 | 0.1000 |
| 1000 | 0.0319 | 0.0316 |
| 10000 | 0.0100 | 0.0100 |

This is a genuine verification rather than an illustration: the theory predicts a specific number at each $n$, and the measurement lands on it. ($\sigma = 1$ for Exponential(1), which makes the predicted column simply $1/\sqrt n$.)

**And the whole proof is three lines.** Linearity gives $E[\bar X_n] = \mu$. Independence makes variances *add*, so $\mathrm{Var}(\bar X_n) = \sigma^2/n$. Chebyshev converts small variance into high probability of being close. That is the entire weak law — and the middle step is the only place independence enters, which is exactly what the next cell exploits.

**The $1/\sqrt n$ is the exchange rate of all empirical work.** To halve your error bar you need **four times** the data. To gain one decimal digit you need **one hundred times**. That is why:

- Monte Carlo integration is expensive to high precision, and why quasi-random and variance-reduction methods exist.
- Benchmark differences of a fraction of a percent need enormous test sets to be meaningful.
- [Hoeffding's](../Concentration/Concentration_Inequalities.ipynb) 38,005-sample requirement for ±1% is the same arithmetic with a confidence level attached.

Every "$\pm$" you have ever written comes from this table.

**Read the $n = 10$ row as the interesting one.** Measured 0.3109 against a predicted 0.3162 — the agreement is *worse* here than at larger $n$, which is expected twice over: the asymptotic argument is weakest at small $n$, and with only 2000 trials the estimate of the standard deviation is itself noisy. Convergence results are asymptotic statements, and checking them at the small end is how you find out where "large $n$" begins for your problem.

**One thing to notice about the hypothesis.** Nothing here required the exponential distribution to be symmetric, bounded, or nice. Finite variance is all the WLLN needs, and the strong law needs only a finite mean. The theorems are cheap to satisfy, which is why they underwrite so much.

### What independence buys — and what breaks without it

The variance-addition step $\mathrm{Var}(\sum X_i) = \sum \mathrm{Var}(X_i)$ *needs* uncorrelatedness. With correlated samples the $1/n$ shrinkage stalls:

In [5]:
# AR(1)-correlated "samples" (ρ = 0.95) vs independent ones: same marginal variance
n = 5000
rho = 0.95
eps = rng.standard_normal((200, n))
ar = np.zeros_like(eps)
for t in range(1, n):
    ar[:, t] = rho * ar[:, t-1] + np.sqrt(1 - rho**2) * eps[:, t]

for name, data in [("independent", eps), ("correlated (ρ=0.95)", ar)]:
    means = data.mean(axis=1)
    print(f"{name:22s} std of X̄_5000 across 200 runs: {means.std():.4f}")
print("→ correlated samples are worth far fewer 'effective' samples —")
print("  the reason time-series folks and MCMC users obsess over effective sample size.")

independent            std of X̄_5000 across 200 runs: 0.0134
correlated (ρ=0.95)    std of X̄_5000 across 200 runs: 0.0835
→ correlated samples are worth far fewer 'effective' samples —
  the reason time-series folks and MCMC users obsess over effective sample size.


**What just happened.** Same number of samples (5000), same marginal variance, same estimator — and the standard error is **6.2× larger** for the correlated stream: 0.0835 against 0.0134.

**Convert that into the number practitioners actually use.** For an AR(1) process the effective sample size is
$$n_{\text{eff}} = n\,\frac{1-\rho}{1+\rho} = 5000 \times \frac{0.05}{1.95} \approx 128.$$
So **5000 correlated samples are worth about 128 independent ones — 2.6% of what you collected.** The predicted standard-error ratio is $\sqrt{5000/128} = 6.24$, against the measured 6.23. Theory confirmed, and the confirmation is what makes the number trustworthy rather than a rule of thumb.

**Where exactly the proof broke.** The WLLN's middle step was $\mathrm{Var}(\sum X_i) = \sum \mathrm{Var}(X_i)$, which needs uncorrelatedness. With correlation the cross-terms survive:
$$\mathrm{Var}\Big(\sum_i X_i\Big) = \sum_i \mathrm{Var}(X_i) + 2\sum_{i<j}\mathrm{Cov}(X_i, X_j),$$
and for positively correlated samples that second sum is large and positive. The $1/n$ shrinkage stalls. Nothing about the estimator changed — the *hypothesis* changed, and one line of the proof stopped holding.

**This is the most common statistical error in applied work, and it always has the same shape.** Treat correlated observations as independent and your error bars come out too small — here by a factor of 6, and by much more at higher $\rho$. Every one of these is the same mistake:

- **MCMC.** Consecutive samples are correlated by construction, with $\rho$ set by the chain's [spectral gap](../Stochastic_Processes/Stochastic_Processes_2.ipynb). A slowly-mixing chain can deliver a million samples worth a few hundred, which is precisely why effective sample size is the standard diagnostic and why reporting raw sample counts is meaningless.
- **Time series.** Randomly shuffled cross-validation on temporally correlated data leaks information between folds and reports optimistic scores.
- **Repeated measures.** Multiple observations of the same user in an A/B test are not independent trials, so naive significance tests over-reject.

**And note the honest framing of what independence buys.** It is not needed for the estimator to be *unbiased* — the correlated average is still centred on the right value, which is why the failure is invisible if you only check the mean. What independence buys is the **rate**. The bias is fine; the error bars are wrong. That is the dangerous kind of failure, because everything looks correct until someone acts on the confidence interval.

**Closing the Analysis track.** ℝ exists → its topology → limits work → sets can be measured → probability is a measure → **averages converge**. Every empirical mean in this curriculum — every training loss, every benchmark, every Monte Carlo estimate — now rests on proved ground, together with an explicit statement of the hypothesis that makes it work and what it costs when that hypothesis fails.

## The Analysis Track, Closed

The arc, start to finish: [ℝ exists](./Real_Number_Systems.ipynb) → [its topology](./Basic_Topology.ipynb) → [limits work](./Numerical_Sequences_and_Series.ipynb) → [sets can be measured](./Measure_Theory.ipynb) → [probability is a measure](./Random_Variables.ipynb) → **averages converge** (this notebook). Every empirical mean in this curriculum now stands on proved ground.

---
## Where next

- [Machine Learning](../../Intro_Mach_Learn/README.md) — training is minimizing an *empirical* average that LLN ties to the true risk.
- [Time Series](../../Intro_Time_Series/README.md) — what changes when independence fails by design.
- [Adaptive Filtering: Kalman](../../Intro_Time_Series/Intro_AdFilt_KF.ipynb) — optimal estimation when you know the correlations.